### DISCUSSION TITLE
How to make an UUID PK generated by String field?

### DISCUSSION THREAD

**Discussion by jinserk** [collerek/ormar#230] [Q&A] [answered]
Hi,

I don't know well about the database, but I'd like to make a table where the entries PK should be UUID, generated from its String field. The string can have varying length so it's not suitable to use it directly as an id. Or at least I need to use any other hash like md5 or sha1 since I want to make the id unique depending only on the string, not from the system or DNS. I'd like to make it id autogenerated once the string is valid from the pydantic's validator.

Is it possible in ormar? Thank you so much!

**Comment by collerek** [comment, marked as answer]
You can use signals for it.

If you register `pre_save` signal you can register a function that will generate uuid from string for you.
You can use uuid3 or uuid5 to generate md5 or sha1 out of string.

Pre_save is triggered before the save happens in db. Check signals docs section.

Something like this.

```python
# model class is your model - one signal can work for several classes - pass a list in that case
@pre_save(ModelClass)
async def before_save(self, sender, instance, **kwargs):
        # string column is a column with your string, pk is a shortcut for primary key column
        instance.pk = uuid.uuid5(instance.string_column)
```

Don't know what exactly is your use case but if you want to ensure that string is stored in a database only once, you can still autogenerate uuids and simply add unique=True constraint on your string column. That way even if you pass a model with different pk but same string you will get an exception.

**Reply by jinserk** [reply]
It's awsome @collerek! Thank you so much!
So the `pre_save` is right after the `validator` of `pydantic` but before `save` to db, am I right? That will be the one I was looking for.
Once the string and its uuid stored in db, I'll make that fields cannot be modified.
Thank you again!

**Reply by collerek** [reply]
Technically validation happens when you either initialize the model or set a value on a model, while the `pre_save` is triggered when you either call `save()` on model or `create()` on queryset. It's triggered before the db save. (After db save there is `post_save`).

I was thinking also about transformations that you can register for model that are run at each assignment/read of values. That would be very similar to signals but registered for a field not a model but run on set/get given column (including init). As signals are `async` and cannot be run from inside `__init__` or `__setattr__` etc. and `default` param on fields cannot accept arguments. That might be a enhancement proposal.

**Reply by jinserk** [reply]
@collerek It is another good advice! I'll try both of them. Thank you once again!

**Reply by collerek** [reply]
Transformations don't exist yet if you mean it by the both of them, that's why it's potential enhancement.

**Reply by jinserk** [reply]
I'm trying to the first suggestion of yours using the `pre_save` signals, but initializing the model object complains as:
```
1 validation error for Mol                                                                                                                                                                                                                                                      
id
  field required (type=value_error.missing)
```

and my code for `Mol` model is:
```
class Mol(ormar.Model):
    class Meta(BaseMeta):
        tablename = "mols"

    id: int = ormar.UUID(primary_key=True, unique=True)
    smiles: str = ormar.String(max_length=256)

    @validator('smiles')
    def make_canonical_smiles(cls, v):
        return utils.standardize_smiles(v)


@ormar.pre_save(Mol)
async def before_save(self, sender, instance, **kwargs):
    # string column is a column with your string, pk is a shortcut for primary key column
    instance.pk = uuid.uuid5(instance.smiles)
```

How to make it work without givening the `id` when I make the `Mol` instance?

**Reply by collerek** [reply]
Primary key is required and has to be unique anyway so no need for unique=True (it's already guaranteed).

Since you cannot make `primary_key` optional you should pass a default function that will generate the id for you on initialization.

You can use `id: int = ormar.UUID(primary_key=True, default=uuid.uuid4)` note no parenthesis. This will generate rundom uuid for you on each initialization that your signal will simply overwrite.

**Reply by jinserk** [reply]
@collerek Thank you so much again!! :heart:

**Reply by jinserk** [reply]
I'm very sorry to bother you again, but it looks like the signal not working as I expected:
```
[ins] In [46]: mols = await Mol.objects.filter(smiles="C=CCC1Oc2cccc(SC)c2-c2ccc3NC(C)(C)C=C(C)c3c21").all()

[ins] In [47]: mols
Out[47]: 
[Mol({'id': UUID('212c0504-c58a-4d0a-b71d-991846610fed'), 'smiles': 'C=CCC1Oc2cccc(SC)c2-c2ccc3NC(C)(C)C=C(C)c3c21'}),
 Mol({'id': UUID('4860a473-7621-4c23-91bc-8910f87ab5d1'), 'smiles': 'C=CCC1Oc2cccc(SC)c2-c2ccc3NC(C)(C)C=C(C)c3c21'})]
```

As you can see the `smiles` str field is the same but the `id` has two different uuids. Here the code I used:

```
class Mol(ormar.Model):
    # fix namespace to generate always the same uuid from smiles
    UUID_NAMESPACE = uuid.UUID('e96f0bba-1c95-4111-90b4-b6d33688667d')

    class Meta(BaseMeta):
        tablename = "mols"

    # default uuid4 is just to avoid an init error.
    # it will be overwritten by pre_save uuid5 from its smiles
    id: str = ormar.UUID(primary_key=True, uuid_format='hex', default=uuid.uuid4)
    smiles: str = ormar.String(max_length=256)

    @validator('smiles')
    def make_canonical_smiles(cls, v):
        return utils.standardize_smiles(v)


@ormar.pre_save(Mol)
async def before_save(self, sender, instance, **kwargs):
    # string column is a column with your string, pk is a shortcut for primary key column
    instance.id = uuid.uuid5(instance.UUID_NAMESPACE, instance.smiles)
```

**Reply by collerek** [reply]
How do you save it and what does utile standardize smiles do?

**Reply by jinserk** [reply]
Here is the storing part:
```
class Sqlite3Backend(StoreBackendBase):
    chunk = 1000

    def __init__(self, *args, **kwargs):
        super().__init__()

    def store(self, data_list: list[dict]):
        invalid = 0
        obj_list = []
        for d in tqdm(data_list, desc="validating", dynamic_ncols=True):
            try:
                obj_list.append(Mol(**d))
            except Exception as e:
                #print(e)
                invalid += 1

        logger.info(f"{invalid} entries are invalid; abandoned")
        asyncio.run(self._async_store(obj_list))

    async def _async_store(self, data_list: list[Mol]):
        total = len(data_list)
        if not total:
            return
        with tqdm(total=total, desc="storing", dynamic_ncols=True) as pbar:
            for s in range(0, total, self.chunk):
                e = min(s + self.chunk, total)
                await Mol.objects.bulk_create(data_list[s:e])
                pbar.update(e - s)

```

and the `utils.standaridze_smiles()` func is just checking the smiles string is chemically valid. It doesn't change the string itself.

**Reply by collerek** [reply]
Ok, so `bulk_create` is all about speed and it does not trigger the signals as it creates one query only that populates the values from dicts so it does not operate on instances (one of the passed arguments in signals is the instance). And this would trigger the signal for each instance anyway so would greatly decrease the speed.

If you gonna populate the values only by initialization (so no instance.string_column = "as") there is a better approach ( for this signal is an over kill anyway🙂).

Simply overwrite Mol init
```
class Mol(ormar.Model):
    # meta etc.
    def __init__(**kwargs):
        kwargs["id"] = uuid.uuid5(namespace, kwargs.get("smiles"))
        super().__init__(**kwargs)
```

Now that this way you can remove the default from us column too.

**Reply by jinserk** [reply]
Thank you for your time to guide me! By the way, I want to make the uuid PK after the smiles validated by `pydantic`'s validator, In that case the uuid generation in `__init__` doesn't look a proper approach...

**Reply by collerek** [reply]
Technically you are right but since you said you do not change string it doesn't matter as the validation happens also in ormar init. So when you call super init if the validation will fail exception will be raised and instance won't be created.

If you really want to do it after validation move super call before the uuid line as after ormar init model will be validated. Note that in that case you need to keep the default as otherwise validation in super call will fail.

**Reply by collerek** [reply]
Of course if you change the order don't update kwargs but self.id

**Reply by jinserk** [reply]
Yep. that's my fault. the `utils.standardize_smiles()` func can return `None` if the given `smiles` string is chemically invalid. 

I'm testing this code:
```
class Mol(ormar.Model):
    # fix namespace to generate always the same uuid from smiles
    _UUID_NAMESPACE: ClassVar[uuid.UUID] = uuid.UUID('e96f0bba-1c95-4111-90b4-b6d33688667d')

    class Meta(BaseMeta):
        tablename = "mols"

    id: str = ormar.UUID(primary_key=True, uuid_format='hex')
    smiles: str = ormar.String(nullable=False, max_length=256)

    def __init__(self, **kwargs):
        kwargs['smiles'] = utils.standardize_smiles(kwargs.get('smiles'))
        kwargs['id'] = uuid.uuid5(self._UUID_NAMESPACE, kwargs['smiles'])
        super().__init__(**kwargs)
```
inspired by your last advice. I guess this could be the best workaround in my use. Still it blocks the `bulk_create` if there is duplicated id but that could be something what I expected. 
Thank you very much again for all your help. You're the super hero!

**Reply by collerek** [reply]
Happy to help :)

Since you set smiles to `nullable=False` if your validator return `None` it should still fail with missing required value, so line `kwargs['smiles'] = utils.standardize_smiles(kwargs.get('smiles'))` should not be required for proper working.

if you want to proceed with bulk_create even with duplicates (I mean when there is a duplicate do not save but do not fail you need one additional query.

Something like:
```python
async def _async_store(self, data_list: list[Mol]):
        total = len(data_list)
        if not total:
            return
        with tqdm(total=total, desc="storing", dynamic_ncols=True) as pbar:
            for s in range(0, total, self.chunk):
                e = min(s + self.chunk, total)
                # get existing ids - in loop to fetch after each chunk for always up to date
                existing_ids = await Mol.objects.values_list("id", flatten=True)
                # skip those that exists
                to_create = [mol for mol in data_list[s:e] if mol.id not in existing_ids]
                await Mol.objects.bulk_create(to_create)
                pbar.update(e - s)
```

**Reply by collerek** [reply]
Alternatively you can fetch the ids once and update it from your data_list in each loop, since you know the ids before saving (much more optimized and quicker) so do not require the response from db

```python
async def _async_store(self, data_list: list[Mol]):
        total = len(data_list)
        if not total:
            return
        with tqdm(total=total, desc="storing", dynamic_ncols=True) as pbar:
            existing_ids = {*await Mol.objects.values_list("id", flatten=True)}        # convert to set
            for s in range(0, total, self.chunk):
                e = min(s + self.chunk, total)
                # skip those that exists
                to_create = [mol for mol in data_list[s:e] if mol.id not in existing_ids]
                existing_ids = existing_ids.union({mol.id for mol in to_create}) # update without db call
                await Mol.objects.bulk_create(to_create)
                pbar.update(e - s)
```

**Reply by jinserk** [reply]
Wow it looks awesome! :+1: :heart:

**Reply by collerek** [reply]
And there is a bug pass of course `to_create` to bulk_create not data_list slice :D

EDIT: FIxed, that what you get from writing code in github :D

**Reply by jinserk** [reply]
Thank you so much again!!!

**Reply by jinserk** [reply]
Hi @collerek,

Sorry to ask a question again, but I'm very confused since:
```
[nav] In [37]: m = await Mol.objects.get(id=uuid.UUID("c07b5e2c-851f-5cea-9083-f908e981087c"))

[ins] In [38]: m
Out[38]: Mol({'id': UUID('13c21807-67f4-5bf2-ba14-bd41e3701c5c'), 'smiles': 'COc1cc(NS(C)(=O)=O)ccc1Nc1c2ccccc2nc2cc(Cl)ccc12'})
```

As you can see, I wanted to get a Mol objects with the id of `c07b5e2c-851f-5cea-9083-f908e981087c` but got a different objects with the id `13c21807-67f4-5bf2-ba14-bd41e3701c5c`. Is it possible?

EDIT: please forget this. I modified the `utils.standardize_smiles` function a little bit and it looks making this confusion. It's still very unclear but let me figure out a little bit more! Sorry for this.

**Reply by jinserk** [reply]
@collerk, what if the result of `utils.standize_smiles()` can be changed?
I put some routines in the function to fix the string and return if the smiles string is invalid. As the result, the id and the smiles fields can be changed together. In that case, I guess the calling of the func in `__init__()` is not working anymore.

Thank you!

**Reply by collerek** [reply]
Ok but I assume you still change it in init or after initialization?

Cause if in init you can change the order and should work.

```python
class Mol(ormar.Model):
    # fix namespace to generate always the same uuid from smiles
    _UUID_NAMESPACE: ClassVar[uuid.UUID] = uuid.UUID('e96f0bba-1c95-4111-90b4-b6d33688667d')

    class Meta(BaseMeta):
        tablename = "mols"

    id: str = ormar.UUID(primary_key=True, uuid_format='hex')
    smiles: str = ormar.String(nullable=False, max_length=256)

    def __init__(self, **kwargs):
        if kwargs.get("id", None) is None:  # set random id here if empty to avoid init problems
            kwargs['id'] = uuid.uuid4()  # alternatively use default, overwrite only if id is not already set
        super().__init__(**kwargs)  # here the validation happens so if the string changes self.smiles will be changed already
        if self.id is None:  # overwrite if not set (new instance, not coming from db) remove if use default
            self.id = uuid.uuid5(self._UUID_NAMESPACE, self.smiles)
```

**Reply by jinserk** [reply]
Thank you again! I guess it's better to cover the case when the field to be changed as:
```
class Mol(ormar.Model):
    # fixed namespace to generate always unique uuid from the smiles
    _UUID_NAMESPACE: ClassVar[uuid.UUID] = uuid.UUID('12345678-abcd-1234-abcd-123456789abc')

    class Meta(BaseMeta):
        tablename = "mols"

    smiles: str = ormar.String(nullable=False, unique=True, max_length=256)
    id: str = ormar.UUID(primary_key=True, index=True, uuid_format='hex')

    def __init__(self, smiles: str, *args, **kwargs):
        canon_smiles = utils.standardize_smiles(smiles)
        id = self.uuid(canon_smiles)
        super().__init__(id=id, smiles=canon_smiles)

    @root_validator()
    def make_canonical_smiles_and_uuid(cls, values):
        values['smiles'] = utils.standardize_smiles(values['smiles'])
        values['id'] = cls.uuid(values['smiles'])
        return values

    @classmethod
    def uuid(cls, smiles):
        return uuid.uuid5(cls._UUID_NAMESPACE, smiles)
```
Here I generated the id in the `__init__` to avoid init error, and also used a `root_validator` of `pydantic` to cover the later change of smiles string. At lease I can cover the smile string update and resulting the id update. Of course there remains a risk to change id only, but in that case the `root_validator` will roll back the id update to maintain the consistency between the two fields. This is what I found as the best so far. I'm leaving this for the other's reference. 

Thank you for all your help and discussion, @collerek!

**Reply by collerek** [reply]
Yeah that should work too in most cases.

Note that ormar also initialize model for you when you read data from database and it does so through `__init__` call too.

So in your case you are recreating the id again also for the models that already have it and my if statements were targeted to avoid it.

Note also that you **really should** pass kwargs to super init as ormar pass special internal parameters that way (i.e. `__pk_only__` flag for relation you didn't select or `__exclude__` to handle fields that should be excluded but had default values when you select subset of field). When parent accept variable args and kwargs you should always pass them just in case.

Note that id can be in kwargs so you need to pop it `kwargs.pop("id", None)` of you implement it.

**Reply by jinserk** [reply]
Aha.. got it. With ormar we'd better to use `kwargs` instead of args then. Thank you for the valuable comment!

**Reply by jinserk** [reply]
I fixed my `Mol` model as follows:
```
class Mol(ormar.Model):
    # fixed namespace to generate always unique uuid from the smiles
    _UUID_NAMESPACE: ClassVar[uuid.UUID] = uuid.UUID('12345678-abcd-1234-abcd-123456789abc')

    class Meta(BaseMeta):
        tablename = "mols"

    id: str = ormar.UUID(primary_key=True, index=True, uuid_format='hex')
    smiles: str = ormar.String(nullable=False, unique=True, max_length=256)

    def __init__(self, **kwargs):
        kwargs['id'], kwargs['smiles'] = self.uuid(kwargs['smiles'])
        super().__init__(**kwargs)

    @root_validator()
    def make_canonical_smiles_and_uuid(cls, values):
        values['id'], values['smiles'] = cls.uuid(values['smiles'])
        return values

    @classmethod
    def uuid(cls, smiles):
        canon_smiles = utils.standardize_smiles(smiles)
        id = uuid.uuid5(cls._UUID_NAMESPACE, canon_smiles)
        return id, canon_smiles
```

But with this `__init__` it complains there exists "unknown" args like:
```
Unknown field 'key' for model Mol
```
Is it intended in the design of `ormar.Model`? How to avoid this error?

EDIT: oops. it's my fault. I input additional keys in dict. Sorry for bothering you.

**Comment by jinserk** [comment]
Hmm.. I found that when I use `Mol.objects.create` without `id`, even though I have the routine to generate the missing `id` in `Mol.__init__()`, it gives me an error as:
```python3
self = <Connection(Thread-2, started 140397017454336)>

    def run(self) -> None:
        """
        Execute function calls on a separate thread.
     
        :meta private:
        """
        while True:
            # Continues running until all queue items are processed,
            # even after connection is closed (so we can finalize all
            # futures)
            try:
                future, function = self._tx.get(timeout=0.1)
            except Empty:
                if self._running:
                    continue
                break
            try:
                LOG.debug("executing %s", function)
>               result = function()
E               sqlite3.IntegrityError: NOT NULL constraint failed: mols.id

../../../.pyenv/versions/3.9.5/envs/kai/lib/python3.9/site-packages/aiosqlite/core.py:102: IntegrityError
```

It looks like `aiosqlite` access is earlier than the `pydantic`s `validator` when I use `create`. This means, that I have to use 
```
mol = Mol(smiles="...")
mol.save()
```
to use `__init__()`?

**Reply by collerek** [reply]
Yes, the result of init is not used with create, ormar sets default on fields with default from the fields dictionary. 

That should probably be marked as a bug (not strictly one but a good practice would be to use init if user overwrite it as you do), can you create an issue for it please? 

As a workaround yes you have to use save but it can be a one call like 
```python
await Mol(**kwargs).save()
```

The performance should be more or less the same with both methods.

**Reply by jinserk** [reply]
Of course, thank you so much @collerek!

**Comment by collerek** [comment]
Fixed in 0.10.12 (via #251)

**Reply by jinserk** [reply]
I tested and it works like charm! Thank you so much! :)
